scheduler/propertyset.go

```diff
@@ -0,0 +1,265 @@
+// values for new allocation placements.
+func NewPropertySet(ctx Context, job *structs.Job) *propertySet {
+	p := &propertySet{
+		ctx:            ctx,
+		jobID:          job.ID,
+		existingValues: make(map[string]struct{}),
+	}
+
+	return p
+}
+
+// SetJobConstraint is used to parameterize the property set for a
+// distinct_property constraint set at the job level.
+func (p *propertySet) SetJobConstraint(constraint *structs.Constraint) {
+	// Store the constraint
+	p.constraint = constraint
+	p.populateExisting(constraint)
+}
+
+// SetTGConstraint is used to parameterize the property set for a
+// distinct_property constraint set at the task group level. The inputs are the
+// constraint and the task group name.
+func (p *propertySet) SetTGConstraint(constraint *structs.Constraint, taskGroup string) {
+	// Store that this is for a task group
+	p.taskGroup = taskGroup
+
+	// Store the constraint
+	p.constraint = constraint
+
+	p.populateExisting(constraint)
+}
+
+// populateExisting is a helper shared when setting the constraint to populate
+// the existing values.
+func (p *propertySet) populateExisting(constraint *structs.Constraint) {
+	// Retrieve all previously placed allocations
+	ws := memdb.NewWatchSet()
+	allocs, err := p.ctx.State().AllocsByJob(ws, p.jobID, false)
+	if err != nil {
+		p.errorBuilding = fmt.Errorf("failed to get job's allocations: %v", err)
+		p.ctx.Logger().Printf("[ERR] scheduler.dynamic-constraint: %v", p.errorBuilding)
+		return
+	}
+
+	// Filter to the correct set of allocs
+	allocs = p.filterAllocs(allocs)
+
+	// Get all the nodes that have been used by the allocs
+	nodes, err := p.buildNodeMap(allocs)
+	if err != nil {
+		p.errorBuilding = err
+		p.ctx.Logger().Printf("[ERR] scheduler.dynamic-constraint: %v", err)
+		return
+	}
+
+	for _, alloc := range allocs {
+		nProperty, ok := p.getProperty(nodes[alloc.NodeID], constraint.LTarget)
+		if !ok {
+			continue
+		}
+
+		p.existingValues[nProperty] = struct{}{}
+	}
+}
+
+// PopulateProposed populates the proposed values and recomputes any cleared
+// value. It should be called whenever the plan is updated to ensure correct
+// results when checking an option.
+func (p *propertySet) PopulateProposed() {
+
+	// Reset the proposed properties
+	p.proposedValues = make(map[string]struct{})
+	p.clearedValues = make(map[string]struct{})
+
+	// Gather the set of proposed stops.
+	var stopping []*structs.Allocation
+	for _, updates := range p.ctx.Plan().NodeUpdate {
+		stopping = append(stopping, updates...)
+	}
+	stopping = p.filterAllocs(stopping)
+
+	// Gather the proposed allocations
+	var proposed []*structs.Allocation
+	for _, pallocs := range p.ctx.Plan().NodeAllocation {
+		proposed = append(proposed, pallocs...)
+	}
+	proposed = p.filterAllocs(proposed)
+
+	// Get the used nodes
+	both := make([]*structs.Allocation, 0, len(stopping)+len(proposed))
+	both = append(both, stopping...)
+	both = append(both, proposed...)
+	nodes, err := p.buildNodeMap(both)
```

Comment:
I think there should be a helper that takes `[]*Allocation` and `map[string]struct{}` and does both the node map building, and property setting. It seems that pattern exists a few times here.